webui/react/src/shared/contexts/stores/UI.tsx

```diff
@@ -0,0 +1,143 @@
+    this.dispatch({ type: StoreActionUI.HideUISpinner });
+  };
+
+  public setMode = (mode: Mode): void => {
+    this.dispatch({ type: StoreActionUI.SetMode, value: mode });
+  };
+
+  public setPageVisibility = (isPageHidden: boolean): void => {
+    this.dispatch({ type: StoreActionUI.SetPageVisibility, value: isPageHidden });
+  };
+
+  public setTheme = (darkLight: DarkLight, theme: Theme): void => {
+    this.dispatch({ type: StoreActionUI.SetTheme, value: { darkLight, theme } });
+  };
+
+  public showChrome = (): void => {
+    this.dispatch({ type: StoreActionUI.ShowUIChrome });
+  };
+
+  public showSpinner = (): void => {
+    this.dispatch({ type: StoreActionUI.ShowUISpinner });
+  };
+}
+
+/**
+ * return a part of the input state that should be updated.
+ * @param state ui state
+ * @param action
+ * @returns
+ */
+const reducerUI = (state: StateUI, action: ActionUI): Partial<StateUI> | void => {
+  switch (action.type) {
+    case StoreActionUI.HideUIChrome:
+      if (!state.showChrome) return;
+      return { showChrome: false };
+    case StoreActionUI.HideUISpinner:
+      if (!state.showSpinner) return;
+      return { showSpinner: false };
+    case StoreActionUI.SetMode:
+      return { mode: action.value };
+    case StoreActionUI.SetPageVisibility:
+      return { isPageHidden: action.value };
+    case StoreActionUI.SetTheme:
+      return {
+        darkLight: action.value.darkLight,
+        theme: action.value.theme,
+      };
+    case StoreActionUI.ShowUIChrome:
+      if (state.showChrome) return;
+      return { showChrome: true };
+    case StoreActionUI.ShowUISpinner:
+      if (state.showSpinner) return;
+      return { showSpinner: true };
+    default:
+      return;
+  }
+};
+const StateContext = React.createContext<StateUI | undefined>(undefined);
+const DispatchContext = React.createContext<Dispatch<ActionUI> | undefined>(undefined);
+
+const reducer = (state: StateUI, action: ActionUI): StateUI => {
+  const newState = reducerUI(state, action);
+  return { ...state, ...newState }; // TODO: check for deep equality here instead of on the full state
+};
+
+const useStore = (): { actions: UIActions; ui: StateUI } => {
+  const context = useContext(StateContext);
+  if (context === undefined) {
+    throw new Error('useStore must be used within a StoreProvider');
+  }
+  const dispatchContext = useContext(DispatchContext);
+  if (dispatchContext === undefined) {
+    throw new Error('useStoreDispatch must be used within a StoreProvider');
+  }
+  return { actions: new UIActions(dispatchContext), ui: context };
+};
+
+export const StoreProvider: React.FC<{ children?: React.ReactNode }> = ({ children }) => {
+  const [state, dispatch] = useReducer(checkDeepEquality(reducer, logger), initUI);
+  return (
+    <StateContext.Provider value={state}>
+      <DispatchContext.Provider value={dispatch}>{children}</DispatchContext.Provider>
+    </StateContext.Provider>
+  );
+};
+
+export default useStore;
```

Comment:
most of the interesting changes are here mainly decision around import and access patterns.
- shifted the default export to be useStore/useUI instead of provider . this should be shorter to typeout to import and it'll be used most frequently but now my ide seems to not be able to auto-suggest an import 🤷🏼‍♂️ 

- naming: sticking with generic names and schemas between multiple potential future useXs `{state, actions} = useUI()` instead of say {ui, uiActions}

- usage pattern: uiActions.xyz or to deconstruct: here I optimized for how I can wrap this up faster.